include/sys/zfs_znode.h

```diff
@@ -199,6 +199,8 @@ typedef struct znode {
 	uint64_t	z_size;		/* file size (cached) */
 	uint64_t	z_pflags;	/* pflags (cached) */
 	uint32_t	z_sync_cnt;	/* synchronous open count */
+	uint32_t	z_sync_writes_cnt; /* synchronous write count */
+	uint32_t	z_async_writes_cnt; /* asynchronous write count */
```

Comment:
To make absolutely certain the accounting is correct, let's go ahead and assert these counters are zero when the znode is destroyed.  The best place to do this would be in the `zfs_znode_cache_destructor()`.

```
ASSERT0(atomic_load_32(&zp->z_sync_writes_cnt));
ASSERT0(atomic_load_32(&zp->z_async_writes_cnt));
```

Reply:
good idea, i was worried about the possibility of the count drifting away from 0 if there are any bugs which would cause all writes to be committed immediately - this should help catch any issue during test runs